### PR TITLE
Add a lint for unused type parameters

### LINKED ISSUE #25871: Should have lint to disallow unused type parameters
The following compiles and runs without warning:

```rust
fn foo<T: Sync> () { }

fn main() {
    foo::<i32>()
}
```

I think that this should at least warn, just like unused imports or unused variables; it would almost always be a result from a mistake in refactoring.

### PR DESCRIPTION
**Pull request by sanxiyn**
Fix #25871.

### CODE REVIEW

File: src/libcore/lib.rs [hunk 1]
```diff
@@ -63,6 +63,7 @@
 #![feature(no_std)]
 #![no_std]
 #![allow(raw_pointer_derive)]
+#![allow(unused_type_parameters)]
```
Comment (huonw) [inline src/libcore/lib.rs]: It seems weird that this is necessary? Where are things unused?

Comment (sanxiyn) [inline src/libcore/lib.rs]: One example is `core::intrinsics::size_of<T>`.

Comment (huonw) [inline src/libcore/lib.rs]: Oh, hm. Shouldn't extern fns with type params consider them used?

**Comment by alexcrichton** [comment]
It looks like there's a number of legitimate use cases for having "unused" parameters, so can this be an allow-by-default lint instead of warn-by-default?

**Comment by sanxiyn** [comment]
The reason this is a lint (and not a hard error, unlike `struct S<T>`) is because there are legitimate use cases. But they are rare, and in my opinion rare enough that writing `#[allow(unused_type_parameters)]` is acceptable.

**Comment by alexcrichton** [comment]
I understand my viewpoint on whether lints should be warn by default is subject, by my believe is that any false positive on a lint means that it should not be warn-by-default. I've found that over time projects end up gaining dozens of little `#[allow]` directives that end up turning into spaghetti because each lint we have has very few false positives but the union of all the false positives means that you're going to have a lot.

**Comment by sanxiyn** [comment]
This is not the first time you argued for allow-by-default. In #10477 you wrote:

> There is a much larger use of `#[allow(dead_code)]` than I would expect throughout the codebase. I think this may be a sign that this should be an allow-by-default lint.

But dead_code lint was accepted warn-by-default. I am curious what changed your mind, because like @huonw on that thread, I think this should be warn-by-default.

**Comment by alexcrichton** [comment]
> This is not the first time you argued for allow-by-default.

I agree! I have historically been very anti-lints as I feel we have far too many already and generate too many warnings for false positives.

> But dead_code lint was accepted warn-by-default. I am curious what changed your mind, because like @huonw on that thread, I think this should be warn-by-default.

Right, but that was almost two years ago. I would want it to be allow-by-default today.

**Comment by nikomatsakis** [comment]
This is an interesting discussion. I wonder if we should adopt the convention that you can write `_T: Foo` to silence the lint, analogous with unused variables?

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to resubmit with a rebase!

**Pull request closed without merging**